### ISSUE TITLE
`DogeForCausalLM` attends to future tokens with the default `sdpa` backend — `SmallDoge/Doge-20M` scores ppl 5 instead of 16, and cached and uncached `generate()` disagree

### ISSUE THREAD

**Issue by blipbyte** [huggingface/transformers#48748]
### System Info

```
- `transformers` version: 5.18.0.dev0
- Platform: Linux-7.0.0-31-generic-x86_64-with-glibc2.39
- Python version: 3.10.20
- Huggingface_hub version: 1.31.0
- Safetensors version: 0.8.0
- Accelerate version: 1.13.0
- Accelerate config: not found
- DeepSpeed version: not installed
- PyTorch version (accelerator?): 2.12.1+cu130 (CUDA)
- Using distributed or parallel set-up in script?: no
- Using GPU in script?: no, the script below is CPU-only
- GPU type: NVIDIA GeForce RTX 3060
```

### Who can help?

@vasqu @Cyrilvallez

### Information

- [x] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

Scoring `SmallDoge/Doge-20M` (the docs-page checkpoint), the perplexity changed with `attn_implementation`. Under the default `sdpa` backend the model is not causal: logits at position *i* depend on tokens after *i*. `eager` is fine.

`main` at `415e6d2f59`, CPU:

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

name = "SmallDoge/Doge-20M"
tok = AutoTokenizer.from_pretrained(name)
text = tok("Paris is the capital and most populous city of France, with an estimated population of over two million residents.", return_tensors="pt")
prompt = tok("The capital of France is", return_tensors="pt")

for impl in ("eager", "sdpa"):
    model = AutoModelForCausalLM.from_pretrained(name, attn_implementation=impl, dtype=torch.float32).eval()
    with torch.no_grad():
        # teacher-forced loss
        loss = model(**text, labels=text["input_ids"]).loss
        # do earlier logits change when only the last token changes?
        other = {k: v.clone() for k, v in text.items()}
        other["input_ids"][0, -1] = 13
        shift = (model(**text).logits[0, :-1] - model(**other).logits[0, :-1]).abs().max()
        # greedy decoding with and without the cache
        cached = tok.decode(model.generate(**prompt, max_new_tokens=12, do_sample=False)[0], skip_special_tokens=True)
        uncached = tok.decode(model.generate(**prompt, max_new_tokens=12, do_sample=False, use_cache=False)[0], skip_special_tokens=True)
    print(f"{impl:5} loss {loss.item():.3f} (ppl {loss.exp().item():.1f}) | earlier logits move by {shift.item():.3f} | cached == uncached: {cached == uncached}")
```

```
eager loss 2.779 (ppl 16.1) | earlier logits move by 0.000 | cached == uncached: True
sdpa  loss 1.613 (ppl 5.0) | earlier logits move by 3.911 | cached == uncached: False
```

A tiny random `DogeConfig` shows the same. It happens whenever `create_causal_mask` takes the sdpa shortcut and returns `None` — unpadded input, an all-ones `attention_mask`, or any cached call — and disappears once a pad token forces the mask to be built (sdpa then matches eager).

Cause: `DogeAttention.prepare_dynamic_mask` folds the causal mask into its dense `dt_states` bias only when handed one (`modeling_doge.py:359-365`) and always passes that bias on as `attention_mask`. Given `None`, the bias reaches sdpa with no causal term, and `is_causal` is off because a mask was supplied.

Not new: the code has been this way since Doge was added (#35891) for any tokenizer mask or cached call; #42367 extended it to the bare call. Nobody has reported it — `has_attentions = False` hides it from the eager-vs-sdpa tests, and the slow `test_Doge_20M_hard` golden pins the sdpa text.

### Expected behavior

`sdpa` and `eager` should agree, as they do once a padding mask is present: a change to token *k* leaves the logits before *k* untouched, and cached and uncached `generate()` agree.

Who this hits: anyone evaluating or fine-tuning Doge with default settings — the loss is too good because the labels are visible — and anyone generating without the cache. The remote-code `modeling_doge.py` in the `Doge-20M` repo always builds the mask.

***I have a fix (one argument at the mask call, the pattern Falcon and the DSA masks use), a regression test and the regenerated golden ready locally — red on `main`, green with the fix. Happy to open the PR if wanted.***

**Comment by shahd-elgouhary** [comment]
Can I take this issue, please?

**Comment by vasqu** [comment]
@blipbyte happy to check a review if you submit a PR 🤗 otherwise I can also check on our side

**Comment by blipbyte** [comment]
Thanks @vasqu — opened #48821 (draft → ready, CI green). The bot assigned other reviewers, so tagging you here in case you want to take it as offered.
